### ISSUE TITLE
IReactToGroupSystem much slower then IReactToEntitySystem and causing GC spikes

### ISSUE THREAD

**Issue by gregee123** [EcsRx/ecsrx.unity#17]
Contrary to what the docs state IReactToGroupSystem is a lot slower than IReactToEntitySystem and also allocates GC memory which is causing spikes easily noticeable in the profiler and practically stopping the game on a moderately fast tablet (galaxy note 2014).

Main culprit is Enumerable.ToArray() which I think is: var entities = accessor.Entities.ToArray(); called during system setup, but could be wrong. Leaking almost 1.0kb per frame is absolutely unacceptable.

Another problem is: IEnumerableExtensions.RunForEach() - it 'only' allocates 32B but obviously it'll quickly accumulate. 

So there are a few separate things here. 

Enumerable.ToArray() should be used sparingly and certainly not every frame which is what ReactToGroup is meant for. 

Perhaps RunForEach() could be called on accessor.Entities directly?? Since it's effectively just an array converted to another array.

In any case it'll still cause some GC allocation as with the way entities are kept in the groupAccessor it's impossible to convert that foreach loop into a normal for loop. So maybe entities should be kept in a plain array[].

![callstack1](https://cloud.githubusercontent.com/assets/7345083/20281180/c4afc434-aaaf-11e6-8e43-83639274e55a.jpg)

I also came across Entity.HasComponents(params Type[] componentTypes) wasting lots of CPU time but I did't drill down into it as I had to find a work-around to get my project working. It is certainly worth looking into.

**Comment by grofit** [comment]
Hey,

I am sure there was a reason why the `ToArray()` was added but off top of my head I cannot remember why so if it performs much better without it and has no change in logic handled I am happy to change it. In general I am more than open to improvements in this area as the main bottleneck in this entire process will be the pool scanning. The logic behind it being faster is that it is just a single subscription that triggers the list to be consumed, rather than the per entity one which has a subscription per entity.

Unfortunately it is not like you can use Tree structures as there is nothing worthwhile to split on like you could with quad/kd structures. One idea which was discussed a while ago was having a smart caching system so basically you would have something (maybe the `GroupAccessor`) listen to the pool changing (i.e entities adding/removing, components adding/removing) and then it would keep an up to date list/array of all the applicable entities. This way although it would take up more memory it would remove the need for the linq scan through the pool every execution.

Anyway like it says on the front page, this was made with functionality in mind and performance took a back seat, but with everything being an interface its super easy to take out one implementation and put a more performant one in if it improves things. So can you check if removing the `ToArray` causes any improvements/issues.

What are your thoughts on the smart caching? as it would still cause an upfront impact on memory (per `GroupAccessor` but group accessors are shared with same groups), but would reduce CPU overhead.

**Comment by gregee123** [comment]
I was about to suggest some sort of reactive caching but it would be a non-trivial change so it wouldn't help me in any way with the way things are currently but if it something you feel like looking into that's definitely going to speed things as I understand currently some systems cause the whole pool of entities to be searched over and over.

FWIW that 'array' in my example only had one entity in it and converting it to another array still took an unbelievable 9.4% on an 3,5 GHz cpu and also left 1kB to clean up per frame.

For now I'd change IEnumerable<IEntity> in GroupAccestor to array then get rid of that ToArray() and changed foreach to a normal for loop.

**Comment by grofit** [comment]
Ideally I would want to keep it as an IEnumerable as that way anyone can use any underlying data structure they want to access/cache the entities as you may have some lazy loading scenarios where systems only want to pull as many instances as a time slice allows. However like I mentioned you should be able to just remove that `ToArray` so it uses the enumerable in the loop.

A for loop is always quicker than a foreach though, just not as nice to work with, but the performance benefits of changing that would be fairly small I think, its also faster to iterate towards 0 rather than away, but again its generally harder to follow.

If you have a demo project to use as a performance baseline I am happy to take a fork of it and have a play around.

**Comment by gregee123** [comment]
As far as foreach vs for it's not about performance it's about garbage on the heap which for doesn't introduce (and which you should never have cause it'll make your game drop fps) but you can't use it with IEnumerable I think.

**Comment by grofit** [comment]
Yeah that's the problem, IEnumerable allows consuming in more complex ways than just iterating, which I would rather keep, as its easy enough for someone who explicitly doesn't want to use IEnumerable to just implement a custom version which removes the foreach and enumerables and uses arrays/lists and for loops. (This could even be done as a plugin which just overrides the default implementations automatically).

I completely agree that the GC overhead is far from ideal, but I don't want to start removing functionality in the sake of optimisation, when the optimisation isn't a requirement but a bonus. This all being said from the crux of your issue its the `ToArray` which is the issue, which if we can focus on that bit specifically to see if it can be removed we can at least change that, as that should not remove functionality or remove use cases.

I will take a look later on in the week around some form of smart caching as that would be one of the biggest performance improvements.

**Comment by gregee123** [comment]
While you're at it you can take a look at how the system reacts to EntityAdded. This I believe is published when an empty entity is created so not sure if there's anything to react to. 

Another thing perhaps event publishing should be inhibited until after the entity is created in a blueprint otherwise there'll be some corner cases where the order in which the components are added in the blueprint will determine which systems get triggered. I'm not sure if that's desirable in a blueprinit.

The easiest thing to accomplish this would be passing the blueprint to the entity (instead of the other way) which could mark itself as under construction while applying the blueprint.

Regardless of details I think this could benefit from some re-engineering.

**Comment by grofit** [comment]
Will make a new issue for you to discuss the event stuff to keep this one focused.

Can you:
1) Try removing the `ToArray` and confirm if it still works and performance improves?
2) Possibly share or isolate and share an example of the major performance issues to use as a baseline?

**Comment by gregee123** [comment]
I can sure do A).

As far as B) on PC this would only be visible in the profiler but it's likely A) will get it fixed and only the small  garbage problem stemming from foreach will remain.

**Comment by gregee123** [comment]
A) took a lil' longer than expected. I did't know that much about the inner workings of linq and so I wan't aware that when in returns IEnumerable<T> it's actually returning a whole chain of operators and iterators which is only evaluated when you actually want to read the data from it. 

So removing ToArray() didn't change much as when foreach call GetEnumarator internally in basically does the same. Slightly faster but no cigar. 

So what I came up with is this small modification of PoolManager (on top of removing that ToArray())

`        public IEnumerable<IEntity> GetEntitiesFor(IGroup group, string poolName = null)
        {
            if(group is EmptyGroup)
            { return new IEntity[0]; }

```
        if (poolName != null)
        { return _pools[poolName].Entities.MatchingGroup(group).ToArray(); } // <<< ToArray()

        return Pools.GetAllEntities().MatchingGroup(group).ToArray(); // <<< ToArray()
    }`
```

This gets rid of the overhead but in turn makes the system not process any matching entities added after that system is setup.

To remedy that you need to add a bit to SystemExecutor.OnEntityComponentAdded() you need to check all groupAccessors (Yes I know they are in pool managers) and if the entity matches any accessors you need to add it to the entities array of that accessor. 

Now there's also a question of removing entities (easy) and removing components (not that hard either)

TL;DR linq + mobile 3d = bad idea. easily puts more strain on CPU than PhysX or at least whatever simple physics unity has.

**Comment by grofit** [comment]
Yeah linq also has some AOT issues on mobiles too, but for devs linq is a great tool and makes it simple to do complex chains. There is always a balance with development simplicity and performance, as if we were to only do what was performant we would basically be writing C++ with smart pointers.

Anyway for the moment then, I will leave this stuff to one side if the performance benefits of removing the `ToArray`   isn't much. While I look into some form of smart caching, as that would hopefully solve both of your issues listed above.

While you wait though if this stuff is actually stopping your game from running just make a plugin:

https://github.com/grofit/ecsrx/tree/master/src/Assets/EcsRx.Examples/PluginExample

Where you just copy and paste removing linq and foreach loops into your own implementations, then when its all updated if you find the performance impact to be lesser than before just disable your plugin.

**Comment by grofit** [comment]
I have put some messages on gitter for you ([November 16, 2016 8:54 AM](https://gitter.im/grofit/ecsrx?at=582c1edde462097a3056cd21)) which details the current state of this, HOPEFULLY these first pass changes should yield some good performance benefits within the GC and CPU time areas.

Once we are both happy we can get this pushed out. 

(Ideally I want you to test this new beta version with your app first to confirm you see same sort of results).
[ecsrx-beta-0.7.0.zip](https://github.com/grofit/ecsrx/files/594519/ecsrx-beta-0.7.0.zip)

**Comment by gregee123** [comment]
So on the one hand the CPU overhead and GC spikes are now gone. However processing is still slower as compared with ReactToEntity at least for small number of entities. So maybe you could add an example scene to EcsRx.Examples with 100-500 rotating cubes to check at what point it actually makes sense to use ReactToGroup instead of ReactToEntity.

**Comment by grofit** [comment]
Can you provide a code example of your `ReactToGroup` method, and `ReactToEntity` methods for both?

```
public IObservable<GroupAccessor> ReactToGroup(GroupAccessor @group)
{
    return // your observable rules here
}
```

I can do a quick sanity test scene later, but I would have assumed if you have N entities in a group and both `ReactToGroup` and `ReactToEntity` are basically the same (i.e every frame do something), the Group accessor should be more performant as it just batches them meaning you ignore a load of function propogation and subscription overhead.

The use cases are mainly separated down to how you expect changes to occur, so for example if you have N entities and you want to react when each of them changes state to "Idle" then you would generally use reaction at the entity level as although the logic is all the same for them, the point when the "change" would occur differs. However with the group based systems it is assumed that they updates would all need to happen at the same time, hence it generally being used for per-frame across all entities systems.

**Comment by gregee123** [comment]
I simply use this: return Observable.EveryUpdate().Select(x => @group); and yes I just replace IEntity with IGroupAccessor.

**Comment by grofit** [comment]
Interesting, I will put together a test scene later somewhere and give it a try, as logically it should be far quicker every frame to have a for loop for each entity and execute on it, than having subscriptions fire as and when they feel like it (for same reason though), unless possibly they are run in some sort of async way which doesnt block the thread as long or something.

**Comment by grofit** [comment]
I have made a quick test scene to run as a test and the metrics I am getting back indicate that the Group is about 20% faster than Entity reactions in systems. I have put a blurb up around it here [November 16, 2016 7:43 PM](https://gitter.im/grofit/ecsrx?at=582cb6f2df5ae966456746bb) let me know your thoughts.

I have added the scene for you to test with if you want, you just need to swap the registered systems over to run each one in isolation to see the results.

[ecsrx.performance.zip](https://github.com/grofit/ecsrx/files/595778/ecsrx.performance.zip)

**Comment by grofit** [comment]
Performance in this area should be greatly improved, and as per metrics in the test scene in the project Groups are faster than Entities for most common scenarios. Closing this.
